### PR TITLE
move to alma9 by default; enable choice through DEFAULT_LINUX_VERSION

### PR DESCRIPTION
**Pull request by h-vetinari**
Alternative to #6548

Closes https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6283
Closes https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6629

We can do this because all the required infrastructure is now in place (c.f. https://github.com/conda-forge/conda-forge.github.io/issues/1941); the last missing piece was to verify how entries of `yum_requirements.txt` match between centos7 & alma8 - the following [overview](https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6283#issuecomment-2440281086) shows that there's a high degree of compatibility (and many things can be replaced by our own build anyway). The handful of differences can be fixed on the affected feedstocks.

Should wait for:
- [x] https://github.com/conda-forge/cdt-builds/pull/71
- [x] https://github.com/conda-forge/docker-images/pull/290 (if we want to add the alma9 images already)
- [x] https://github.com/conda-forge/docker-images/pull/291

### CODE REVIEW

**Comment by h-vetinari** [comment]
For the reasoning behind https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6626/commits/24b7e66bf72d7b150c4221c5fed362d0bd0f98c7, see discussion [here](https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6548#discussion_r1823652519). The preceding commit fails if a feedstock would override `c_stdlib_version`, see also this [branch](https://github.com/h-vetinari/faiss-split-feedstock/commits/pinning_test) where I've tested this for an affected feedstock.

https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6626/commits/24b7e66bf72d7b150c4221c5fed362d0bd0f98c7 would thus be necessary to make the docker image (sufficiently) independent from `c_stdlib_version`, so that they can be set independently (if indeed this is something we desire to enable; one reason why is that the image version doesn't _have_ to affect the sysroot-baseline, in particular if it's only for being able to load a build tool or test dependency).

**Comment by h-vetinari** [comment]
@jakirkham, you mentioned in the core call just now that all the CUDA feedstocks already set `os_version` explicitly? In only did a spot check and it [looks like](https://github.com/conda-forge/cuda-nvcc-impl-feedstock/blob/main/conda-forge.yml) this for example:
```yaml
os_version:
  linux_64: cos7
  linux_aarch64: cos7
  linux_ppc64le: cos7
```
Following https://github.com/conda-forge/docker-images/pull/291 (using the consolidated names that allow choosing the distro version), this would eventually have to move to
```yaml
os_version:
  linux_64: cos7
  linux_aarch64: ubi8
  linux_ppc64le: ubi8
```
or just
```yaml
os_version:
  linux_64: cos7
```
because we don't/won't support distro overrides for aarch/ppc. In either case, once we've confirmed that we got `os_version` set in the relevant feedstocks doing binary repackaging, this PR should be ready. I'll leave out the image renaming stuff for #6687

**Comment by h-vetinari** [comment]
> once we've confirmed that we got `os_version` set in the relevant feedstocks doing binary repackaging, this PR should be ready.

@jakirkham do you have a list of cuda-related feedstocks that I can double-check?

**Comment by h-vetinari** [comment]
Gentle ping @jakirkham

**Comment by jakirkham** [comment]
Thanks Axel! 🙏

Please don't feel obliged to track or make these changes. This is something NVIDIA plans to pick up. In fact folks are eager to adopt these changes once available

There is not one comprehensive list as there are the CUDA Toolkit (CTK) packages, CUDA Math libraries, CUDA DL libraries, etc. Plus more packages are being added

That said, for the CUDA Toolkit packages we handled this work as part of issue ( https://github.com/conda-forge/cuda-feedstock/issues/26 ), this has a list of CTK feedstocks that had this change included

Other NVIDIA feedstocks have been encouraged to follow the same pattern

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -128,36 +126,63 @@ channel_targets:
...
-  - quay.io/condaforge/linux-anvil-ppc64le-cuda:11.8      # [ppc64le and os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
   - quay.io/condaforge/linux-anvil-aarch64-cuda:11.8      # [aarch64 and os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-aarch64"]
+  - quay.io/condaforge/linux-anvil-ppc64le-cuda:11.8      # [ppc64le and os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-ppc64le"]
   # CUDA 11.8 arch: cross-compilation (build != target)
-  - quay.io/condaforge/linux-anvil-cuda:11.8              # [ppc64le and os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-64"]
   - quay.io/condaforge/linux-anvil-cuda:11.8              # [aarch64 and os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-64"]
+  - quay.io/condaforge/linux-anvil-cuda:11.8              # [ppc64le and os.environ.get("CF_CUDA_ENABLED", "False") == "True" and os.environ.get("BUILD_PLATFORM") == "linux-64"]
```
Comment (jakirkham) [inline recipe/conda_build_config.yaml:152]: To make sure I'm not missing something, is there a reason these were reordered?

**Comment by h-vetinari** [comment]
> Please don't feel obliged to track or make these changes. This is something NVIDIA plans to pick up. In fact folks are eager to adopt these changes once available

I agree that this makes sense, but in the last core call the ask had been to commit `os_version: ...` into the binary repackaging feedstocks before switching. That would currently be a no-op, so we could even just push them with a `[ci skip]`. I'm willing to do that (or help), but I don't have a list of feedstocks (well, except perhaps searching for "cuda" in our feedstock names).

File: recipe/conda_build_config.yaml (hunk 1 above)
Comment (h-vetinari) [reply recipe/conda_build_config.yaml:152]: Yeah, I tried to sort everything as `x64` - `aarch64` - `ppc64le`. It's easy to make mistakes when neighbouring blocks have different order. 😅

**Comment by jakirkham** [comment]
The issue link above hopefully answers that question. Admittedly not comprehensively, but that is pretty close

Will look for other cases we may have missed and file issues

**Comment by jakirkham** [comment]
Found a few stragglers. Have put together a list in this comment: https://github.com/conda-forge/cuda-feedstock/issues/28#issuecomment-2487591232

Will start working through those. Also will update the list if we find more

**Review by carterbox — APPROVED** [review]
This is the one that we're going with instead of https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6548, right?

My understanding is that the following is the expected behavior because it is what I observed when using the PR to render a feedstock.

| os_version | container x86 | container aarch64 | container ppc64le |
| ---------- | ------------- | ----------------- | ------------------|
| undefined  | linux-anvil-alma-x86_64:9 | linux-anvil-alma-aarch64:9 | linux-anvil-alma-ppc64le:9
| alma9      | linux-anvil-alma-x86_64:9 | linux-anvil-alma-aarch64:9 | linux-anvil-alma-ppc64le:9
| alma8      | linux-anvil-alma-x86_64:8 | linux-anvil-alma-aarch64:8 | linux-anvil-alma-ppc64le:8
| wrong      | linux-anvil-comp7 | linux-anvil-comp7 | linux-anvil-comp7
| alma7      | linux-anvil-comp7 | linux-anvil-comp7 | linux-anvil-comp7
| cos7       | linux-anvil-cos7-x86_64 | linux-anvil-aarch64 | linux-anvil-ppc64le

Except for CUDA builds where os_versions that are not one of the valid ones (alma9, alma9, cos7) break the zipping of the keys, but this is fine. It's probably a feature that using an invalid os_version doesn't rerender.

CUDA 12 builds follow the non-cuda builds exactly.

CUDA 11 has only one container option for each architecture.

| os_version | container x86 | container aarch64 | container ppc64le |
| ---------- | ------------- | ----------------- | ------------------|
| undefined  | linux-anvil-cuda:11.8 | linux-anvil-aarch64-cuda:11.8 | linux-anvil-ppc64le-cuda:11.8
| alma9      | linux-anvil-cuda:11.8 | linux-anvil-aarch64-cuda:11.8 | linux-anvil-ppc64le-cuda:11.8
| alma8      | linux-anvil-cuda:11.8 | linux-anvil-aarch64-cuda:11.8 | linux-anvil-ppc64le-cuda:11.8
| cos7       | linux-anvil-cuda:11.8 | linux-anvil-aarch64-cuda:11.8 | linux-anvil-ppc64le-cuda:11.8

I expect that `os_version` is completely decoupled from `c_stdlib_version`, so they should not affect each other.

**Comment by h-vetinari** [comment]
Your list is correct as of this PR, but I think it'd actually be better to pull in the changes from https://github.com/conda-forge/docker-images/issues/293 / https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/6687 into this PR. I'll update https://github.com/conda-forge/docker-images/pull/291 to finish the rest of the necessary changes.

I have thought about the fact that only a single value of `os_version` can be used per platform, and how to handle the split between different distros between CUDA 11.8 and the rest. I'll push some changes to this effect momentarily. The result would look as follows (not counting the cross-compilation split for cuda 11.8)

| os_version | container x86 | container aarch64 | container ppc64le | for CUDA 11.8 |
| ---------- | ------------- | ----------------- | ------------------|--|
| undefined  | linux-anvil-x86_64:alma9 | linux-anvil-aarch64:alma9 | linux-anvil-ppc64le:alma9 | no |
| alma9     | linux-anvil-x86_64:alma9 | linux-anvil-aarch64:alma9 | linux-anvil-ppc64le:alma9 | no |
| alma8     | linux-anvil-x86_64:alma8 | linux-anvil-aarch64:alma8 | linux-anvil-ppc64le:alma8 | no |
| ubi8       | linux-anvil-x86_64:alma8 | linux-anvil-aarch64:alma8 | linux-anvil-ppc64le:alma8 | no |
| cos7       | linux-anvil-x86_64:cos7   | linux-anvil-aarch64:cos7    | linux-anvil-ppc64le:cos7 | no |
| undefined  | l.-a.-x86_64-cuda11.8:ubi8 | l.-a.-aarch64-cuda11.8:ubi8 | l.-a.-ppc64le-cuda11.8:ubi8 | yes |
| alma9     | l.-a.-x86_64-cuda11.8:ubi8 | l.-a.-aarch64-cuda11.8:ubi8 | l.-a.-ppc64le-cuda11.8:ubi8 | yes |
| alma8     | l.-a.-x86_64-cuda11.8:ubi8 | l.-a.-aarch64-cuda11.8:ubi8 | l.-a.-ppc64le-cuda11.8:ubi8 | yes |
| ubi8       | l.-a.-x86_64-cuda11.8:ubi8 | l.-a.-aarch64-cuda11.8:ubi8 | l.-a.-ppc64le-cuda11.8:ubi8 | yes |
| cos7       | l-a-x86_64-cuda11.8:cos7   | l.-a.-aarch64-cuda11.8:ubi8 | l.-a.-ppc64le-cuda11.8:ubi8 | yes |

For wrong values of `os_version`, we could add something like the following, but arguably, failed renders are hard to debug, so IMO this should rather be a linter rule in smithy
```yaml
  # intentional breakage for unsupported values of DEFAULT_LINUX_VERSION
  - quay.io/condaforge/this_image_does_not_exist    # [linux and os.environ.get("DEFAULT_LINUX_VERSION", "alma9") not in ("cos7", "ubi8", "alma8", "alma9")]
  - quay.io/condaforge/this_image_does_not_exist    # [linux and os.environ.get("DEFAULT_LINUX_VERSION", "alma9") not in ("cos7", "ubi8", "alma8", "alma9")]
```

**Comment by h-vetinari** [comment]
OK, https://github.com/conda-forge/docker-images/pull/291 was merged and I've tested it on the faiss feedstock with several variations of `cos7` / `ubi8` / `alma8` / `alma9`, and it did the right thing in all those cases (compare table above).

In other words, this PR should be ready. :)

I'll write up the announcement that John [asked](https://github.com/conda-forge/docker-images/issues/293#issuecomment-2487482059) for shortly.

**Comment by h-vetinari** [comment]
> so IMO this should rather be a linter rule in smithy

I started implementing this in smithy, but ran into https://github.com/conda-forge/conda-smithy/issues/2152, which we should very likely solve first.

**Comment by h-vetinari** [comment]
I have a smithy PR for a linter rule now: https://github.com/conda-forge/conda-smithy/pull/2155

**Comment by jakirkham** [comment]
Thanks Axel! 🙏

Opened a couple of test PRs

Was able to re-render ok

Though did see some issues pulling the Docker images. Were the images from PR ( https://github.com/conda-forge/docker-images/pull/291 ) pushed ok? Or are there some naming changes that still need to be picked up here?

* https://github.com/conda-forge/nccl-feedstock/pull/120#issuecomment-2492426414
* https://github.com/conda-forge/zlib-feedstock/pull/87#issuecomment-2492436723

**Comment by h-vetinari** [comment]
Thanks for testing! 🙏 

> Though did see some issues pulling the Docker images. Were the images from PR ( [conda-forge/docker-images#291](https://github.com/conda-forge/docker-images/pull/291) ) pushed ok? Or are there some naming changes that still need to be picked up here?

I had double-checked the CI runs there, and aside from being green, the deploy step looks normal too. Not sure what could be causing the
```
+ docker pull quay.io/condaforge/linux-anvil-x86_64:alma8
Error response from daemon: unauthorized: access to the requested resource is not authorized
```
you're seeing.

**Comment by isuruf** [comment]
Default setting for new images on quay.io is private. Made them public just now.

**Comment by jakirkham** [comment]
Ah right 😅 

Thanks Isuru! 🙏

Remembered there was something on Quay.io that needed an extra tweak. Had forgotten exactly what it was

Restarted the failed CI runs in those test PRs

**Comment by jakirkham** [comment]
Looks like both of those builds passed! 🥳

Also looks like the recent work on changing the `cdt_name` to `conda` in PR ( https://github.com/conda-forge/conda-forge-ci-setup-feedstock/pull/365 ) has fixed CUDA cross-compilation on Alma 8 (am guessing Alma 9 would work similarly)

**Comment by h-vetinari** [comment]
@conda-forge/core, this should be ready to go and I'd welcome reviews. 🙏 

It's been tested both by me and by @jakirkham, and covers what we discussed in the core call (c.f. https://github.com/conda-forge/conda-forge-pinning-feedstock/issues/6283 & https://github.com/conda-forge/docker-images/issues/293)

To accompany this, I also have an [announcement](https://github.com/conda-forge/conda-forge.github.io/pull/2380) (as suggested by John) and a [doc update](https://github.com/conda-forge/conda-forge.github.io/pull/2377) ready. A [linter rule](https://github.com/conda-forge/conda-smithy/pull/2155) for the distrotags in `os_version` is waiting for some dependent work and will follow soon, but we don't need to wait for that here.

**Review by jakirkham — APPROVED** [review]
Looking forward to these improvements! Thanks again Axel for working with us on them 😄

**Pull request merged**